### DISCUSSION TITLE
Get only outer or inner paths/outline from text2path

### DISCUSSION THREAD

**Discussion by jkraehling** [mozman/ezdxf#877] [Q&A] [answered]
Hi all, I was trying to figure out how to convert only the inner or outer outline from a text to path.

I use the function ezdxf.addons.text2path.make_paths_from_str().

At the moment it creates both contours exclusively, but lets say I want to delete only the inner or the outer one, how do I achieve that?

Additionally, it would be desirable if possible to create not only lines from the fonts, but also arcs. I had hoped that by converting to polylines also arcs are created:

```Python
import ezdxf
from ezdxf.addons import text2path
from ezdxf.tools import fonts
from ezdxf.enums import TextEntityAlignment



doc = ezdxf.new()
msp = doc.modelspace()

paths = text2path.make_paths_from_str(
    "Text Example Ø123",
    font=fonts.FontFace(family="Arial"),
    size=4,
    align=TextEntityAlignment.LEFT,
    length=0
)

polylines = ezdxf.path.to_polylines2d(paths, segments=2, distance=0.01)
for pl in polylines:
    msp.add_entity(pl)
    pl.explode()

doc.saveas("Text Example Ø123.dxf")
```

**Comment by mozman** [comment, marked as answer]
This is a simple but not bulletproof way to archive this. Test that the center of the path's bounding box is inside the previous path's bounding box:

```Python
from ezdxf.math import BoundingBox2d

def outer_paths(paths):
    prev = None
    for path in paths:
        if prev is None:
            prev = BoundingBox2d(path.control_vertices())
            yield path
            continue
        bbox = BoundingBox2d(path.control_vertices())
        if not prev.inside(bbox.center):
            yield path
        prev = bbox


def inner_paths(paths):
    prev = None
    for path in paths:
        if prev is None:
            prev = BoundingBox2d(path.control_vertices())
            continue
        bbox = BoundingBox2d(path.control_vertices())
        if prev.inside(bbox.center):
            yield path
        prev = bbox


polylines = ezdxf.path.to_polylines2d(outer_paths(paths), segments=2, distance=0.01)
for pl in polylines:
    msp.add_entity(pl)
    pl.explode()

doc.saveas("Text Example Ø123.dxf")
```

outer paths:
![outer_paths](https://github.com/mozman/ezdxf/assets/4230966/02f4f8e3-c6b3-47d0-bb93-54234a15023d)


inner paths:
![inner_paths](https://github.com/mozman/ezdxf/assets/4230966/16fb070e-3fb3-46ec-9fc0-3b94123f6de1)

**Reply by jkraehling** [reply]
Thank you for the quick answer, works pretty good!

**Reply by changkeke** [reply]
That's good. How can I transfer these contours with inner holes to Geojson's MutiPolygon?

**Reply by mozman** [reply]
https://ezdxf.mozman.at/docs/addons/geo.html

1. convert paths to HATCH entities
2. covert HATCH entities to GeoJSON proxy instances
3. export JSON data

```Python
# untested code may contain errors!!!!
import json
from ezdxf import path
from ezdxf.addons import geo

hatches = path.to_hatches(paths, edge_path=False)  # GeoJSON does not support curves
geo_proxy = geo.proxy(hatches)

with open("my_geo_json_data.json", "wt", encoding="utf8") as fp:
        json.dump(geo_proxy.__geo_interface__, fp, indent=2)

```

GeoJSON is just JSON and not that hard to create: https://pypi.org/project/geojson/

The important part is to convert a `Path` instance to a `LineString` or `Polygon`: `path.flattening(distance=your_required_precision)`

**Reply by changkeke** [reply]
Thanks! I have just tried the addons-geo interface and have the following problems:
1. Hatch transfered to Geojson's Polygon doesn't resolve the contour-with-inner-holes's feature well, I think the MultiPolygon will do better;
2. An question:  Doesn't geo-interface deal with the texture of the HATCH pattern and the complex linetype of LINE,LWPOLYLINE...? (Because, I'm now trying to transfer the dxf file to geojson format files almost completely, also with each entity's style will be transferred in the Geojson's 'feature.properties' attribute....)
